CODEX-MARKETPLACE:  RFC Feedback

jimstir opened the pull request:
Made should changes to the marketplace RFC.

jimstir commented on specs/marketplace.md:
```diff
@@ -116,9 +89,22 @@ A client has two primary responsibilities:
+To create a request to persist a dataset on the Codex network, client nodes MUST split the dataset into data chunks, $(c_1, c_2, c_3, \ldots, c_{n})$.
+The data chunks are then encoded with the erasure coding method and the provided input parameters.
+
+The applied erasure coding method, used on data being persisted all SPs,
+SHOULD be the [Reed-Solomon algorithm](https://hackmd.io/FB58eZQoTNm-dnhu0Y1XnA).
+If the client node uses a different algorithm, once storage request is made,
+SP will not be able to submit valid proofs.
+
```
Bring this to discussion once more because I am still not sure it should be MUST here. Does the smart contract reject a request when data is not erasure encoded, or do SP check when data is downloaded? I believe the `MUST` is required when the validation occurs, which is mentioned later in the spec.

jimstir replied:
Ok, will keep at `MUST`.  I also changed the last statement to include " When dataset is encoded with different algorithm, the dataset will be inaccessible."

AuHau replied:
Sounds good - thanks

AuHau reviewed (changes requested):
Some comments that needs to be fixed.

I noticed one thing. We use the phrase "client node(s)" - eq. the role name with the term "node", but for SPs we often just leave it as "SP", which imho is not nice. I guess this was introduced by Marcin and I have missed it. Could you please go over the spec and fix "SP" to "SP node(s)"?

jimstir commented on specs/marketplace.md:
```diff
@@ -116,9 +89,22 @@ A client has two primary responsibilities:
+The CID for the `Manifest` MUST then be used as the `cid` for the stored dataset.
 
-After the dataset is prepared, a client node MUST call the smart contract function `requestStorage(request)`, providing the desired request parameters in the `request` parameter. The `request` parameter is of type `Request`:
+After the dataset is prepared,
+a client node MUST call the smart contract function `requestStorage(request)`,
```
This doesn't change the document, just following [sembr](https://sembr.org/) for easier reading when editing. In the future, the entire document will follow sembr, but is not enforced in this repo.

AuHau replied:
Hmm cool concept the `sembr` did not know about it. Alright, lets keep it.

jimstir commented on specs/marketplace.md:
```diff
@@ -152,27 +138,30 @@ The the table below provides the description of the `Request` and the associated
-| `expiry` | `uint256` | Timeout in seconds during which all the slots have to be filled, otherwise Request will get cancelled. The final deadline timestamp is calculated at the moment the transaction is mined. |
+| `expiry` | `uint256` | Timeout in seconds during which all the slots have to be filled, otherwise request will get cancelled. The final deadline timestamp is calculated at the moment the transaction is mined. |
 | `nonce` | `byte32` | Random value to differentiate from other requests of same parameters. It SHOULD be a random byte array. |
-| `reward` | `uint256` | Amount of tokens that will be awarded to SPs for finishing the storage request. It MUST be an amount of Tokens offered per slot per second. The Ethereum address that submits the `requestStorage()` transaction MUST have [approval](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#IERC20-approve-address-uint256-) for the transfer of at least an equivalent amount i ...
```
This is obvious, I do not think it should be capitalized.

AuHau replied:
Okey, sounds good.

emizzle replied:
I agree this should be capitalised because "Tokens" is a proper noun in this case. We are using it in lieu of "CDX" or "Codex tokens" since this is not legally approved, and may change in the future.

AuHau replied:
So you are saying keep it uppercased? Then lets revert the lower-casing of the Definition table as well?

emizzle replied:
Hmm, I really don't know what the orthodoxy is with a definitions table in a spec. My comment refers to the capitalization of the word "Tokens" in a sentence. In which case, it is *generally* (but not always) a proper noun.

AuHau replied:
I asked Jimmy to lowercase all the "definitions" in the Definition table because of lowercasing of the "Token": https://github.com/codex-storage/codex-spec/pull/7/files#r1785680631

jimstir replied:
@emizzle This decision is really a matter of preference. I suggested lowercasing because the definition defines an ERC-20 token used with Codex and is the only token type mentioned in the spec. This makes the mention of "token" to be a generic ERC-20 token.  This spec does not introduce or define a "CDX/Codex token".  In this context, tokens should be capitalized and the Codex token should be defined (contract address, symbol).

jimstir commented on specs/marketplace.md:
```diff
@@ -34,73 +45,35 @@ The keywords “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL N
+“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in [2119](https://www.ietf.org/rfc/rfc2119.txt).
 
-A node that wishes to participate in the Codex persistence layer MUST implement one or more roles described in this document.
+The marketplace contract manages storage requests,
+maintains the state of allocated storage slots, and orchestrates SP rewards,
+collaterals, and storage proofs.
+A node that wishes to participate in the Codex persistence layer MUST implement one or
+more roles described in this document.
```
Movtivation section does not belong in the semantic or specification section because it is does not have details about the protocol. More like an introduction to the topic.

jimstir commented on specs/marketplace.md:
```diff
@@ -194,32 +183,43 @@ The following tasks need to be considered when hosting a slot:
 - `requestId` - the ID of the request.
 - `slotIndex` - the slot index that the node wants to fill.
 - `proof` - the `Groth16Proof` proof structure, generated over the slot data.
+- `collateral` - The amount of tokens required to fill a slot
```
Correct, I looked at the wrong function in the codex nim implementation instead of eth contracts implementation. 👍

emizzle replied:
Please don't forget to update this 👍

AuHau replied:
I have merged this in myself, that should be already included.

emizzle replied:
Ah ok, in master I assume?

AuHau replied:
Sorry no, I meant "committed the suggestion" myself. I don't see the `collateral` parameter in the listing anymore - you do?

AuHau replied:
It has been removed here: https://github.com/codex-storage/codex-spec/pull/7/files#diff-3ada29a68d58e7d99e24f0b550cd3b2ffbe6cc2b613077071b0d6cc2beff6c4cR245

jimstir commented on specs/marketplace.md:
```diff
@@ -246,7 +247,50 @@ The repair process proceeds as follows:
-4. The SP MUST call the `fillSlot()` smart contract function with the same parameters and collateral allowance as described in the [Filling Slots](#filling-slot) section.
+4. The SP MUST call the `fillSlot()` smart contract function with the same parameters and collateral allowance as described in the [Filling Slots](#filling-slots) section.
+
+Below is a diagram depicting the proving and repair process:
```
I seems like the two diagrams explain the request state a little different. The one I moved mentions the repair process, so I thought it would be better to the repair section. I moved it again, I believe it could be added to the filling slots section.

jimstir commented:
> Some comments that needs to be fixed.
> 
> I noticed one thing. We use the phrase "client node(s)" - eq. the role name with the term "node", but for SPs we often just leave it as "SP", which imho is not nice. I guess this was introduced by Marcin and I have missed it. Could you please go over the spec and fix "SP" to "SP node(s)"?

Updated "SP" terms to "SP node" and "SPs" to "SP nodes" except for within the diagram. Also moved definitions back in Semantics section.

AuHau reviewed (changes requested):
Jimmy, I have made earlier few suggestions for you to include in the PR. Some of them are even blockers (like the `collateral` parameter). It is great that I guess you "agree" with them (eq. giving them thumbs-up) but then why you didn't include them in the PR?! Why then are you asking for re-review if they are not included? I mean I will "Request changes" again and will have to wait again for you to incorporate those changes and doing another "re-review" afterward... 

Also, please fix the conflicts with master.

jimstir commented on specs/marketplace.md:
```diff
@@ -43,30 +55,19 @@ more roles to maintain a reliable persistence layer for users.
-collaterals, and storage proofs.
-A node that wishes to participate in the Codex persistence layer MUST implement one or
-more roles described in this document.
+| token               | ERC20-based token used within the Codex network.     |
```
Sorry for missing a few of the feedback in the re-review. I updated the other terms in the definition table, added the missing statement to collateral description. Also updated the smart contract definition to remove the word "smart contract" with "set of rules". Just removing smart contract wording in the description.

AuHau commented:
Great! Thanks and LGTM now.

AuHau reviewed (approved):
Jimmy, you still left out one and actually the most important suggestion I had for you - the one about adding "collateral" parameter to `fillSlot`. I just couldn't find enough patience in me anymore and committed it just so we can be over with this.

emizzle reviewed (commented):
There were a few comments where the conversation was marked as "resolved", but the request was not actually resolved. I went through and marked the unresolved comments as "unresolved".

Also, the description of changes in this PR is lacking details. Could you please update it to list the changes that were made, please?

AuHau commented on specs/marketplace.md:
```diff
@@ -152,27 +138,30 @@ The the table below provides the description of the `Request` and the associated
-| `reward` | `uint256` | Amount of tokens that will be awarded to SPs for finishing the storage request. It MUST be an amount of Tokens offered per slot per second. The Ethereum address that submits the `requestStorage()` transaction MUST have [approval](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#IERC20-approve-address-uint256-) for the transfer of at least an equivalent amount i ...
-| `collateral` | `uint256` | The amount of tokens that SPs submit when they fill slots. Collateral is then slashed or forfeited if SPs fail to provide the service requested by the storage request (more information in the [Slashing](#Slashing) section). |
+| `reward` | `uint256` | Amount of tokens that will be awarded to SPs for finishing the storage request. It MUST be an amount of tokens offered per slot per second. The Ethereum address that submits the `requestStorage()` transaction MUST have [approval](https://docs.openzeppelin.com/contracts/2.x/api/token/erc20#IERC20-approve-address-uint256-) for the transfer of at least an equivalent amount i ...
+| `collateral` | `uint256` | The amount of tokens that SPs MUST submit when they fill slots. Collateral is then slashed or forfeited if SPs fail to provide the service requested by the storage request (more information in the [Slashing](#slashing) section). |
```
This has been already incorporated.

AuHau commented:
> There were a few comments where the conversation was marked as "resolved", but the request was not actually resolved. I went through and marked the unresolved comments as "unresolved".

Hmm I think most of those are actually resolved except one.

emizzle commented:
> Hmm I think most of those are actually resolved except one.

No problem. I didn't know there were upstream changes.
